tests/model_notebooks/test_notebooks.py

```diff
@@ -0,0 +1,361 @@
+
+        Raises:
+            RuntimeError: If kernel cannot be registered
+        """
+        if not self._check_kernel_exists():
+            logger.info(f"Kernel '{self.kernel_name}' not found, attempting to register...")
+
+            if not self._register_kernel():
+                raise RuntimeError(
+                    f"Failed to register kernel '{self.kernel_name}'. "
+                    f"Ensure conda environment '{self.conda_env_name}' exists and "
+                    f"has ipykernel installed."
+                )
+
+    def _discover_notebook_paths(self) -> List[Path]:
+        """
+        Discover all Jupyter notebook files in the notebook directory.
+
+        Searches only the top level of the directory (non-recursive) and filters
+        out checkpoint files.
+
+        Returns:
+            List of Path objects for discovered notebooks
+        """
+        notebooks = [
+            nb
+            for nb in self.notebook_directory.glob("*.ipynb")
+            if ".ipynb_checkpoints" not in str(nb)
+        ]
+
+        logger.info(f"Found {len(notebooks)} notebook(s) in {self.notebook_directory}")
+
+        return sorted(notebooks)
+
+    def _execute_notebook(self, notebook_path: Path) -> bool:
+        """
+        Execute a single notebook using papermill.
+
+        Catches all exceptions including SystemExit and KeyboardInterrupt to prevent
+        pytest from exiting with code 2 (which would kill the SLURM session).
+        All failures are converted to regular test failures (AssertionError).
+
+        Args:
+            notebook_path: Path to the notebook to execute
+
+        Returns:
+            True if execution succeeded, False if it failed
+        """
+        try:
+            logger.info(f"Running notebook {notebook_path.name}...")
+            logger.info(f"Using kernel: {self.kernel_name}")
+
+            # Execute notebook with papermill
+            # The kernel is automatically registered in __init__ if not already present
+            pm.execute_notebook(
+                input_path=str(notebook_path),
+                output_path=str(notebook_path),
+                parameters={"model_dir": str(self.model_dir)},
+                kernel_name=self.kernel_name,
+                execution_timeout=self.timeout,
+            )
+
+            logger.success(f"✓ {notebook_path.name} completed successfully")
+
+            return True
+
+        except BaseException as e:
+            # Catch ALL exceptions including SystemExit and KeyboardInterrupt
+            # to prevent pytest exit code 2 which kills the SLURM session.
+            # This ensures test failures (exit code 1) instead of interruptions (exit code 2).
+            error_msg = f"{type(e).__name__}: {str(e)}" if str(e) else type(e).__name__
+            logger.error(f"✗ {notebook_path.name} failed with error: {error_msg}")
+            self.failed_notebooks[notebook_path] = e
```

Comment:
Where do you use the errors you are saving here? If you're not using them, it's more natural to add the notebook to the collection of failed notebooks in `test_run_notebooks()` like you do for the successes.

Reply:
This is adding the notebook path to the dict of failed notebooks and the value is the exception. We currently do not log the exception to the user but I can do that in a follow up pr.